---
title: "Lab 14: RNA-seq mini project"
author: "Torrey Rhyne (A14397504)"
format: pdf
---

## Workflow
1. DESeq analysis
2. Pathway analysis
3. Gene ontology 
4. Reactome analysis


```{r}
library(DESeq2)
```

## Import data

```{r}
# import
metaFile <- "GSE37704_metadata.csv"
colData = read.csv(metaFile, row.names=1)
countFile <- "GSE37704_featurecounts.csv"
countData = read.csv(countFile, row.names=1)

# take a look
head(colData)
head(countData)
```

Q. Complete the code below to remove the troublesome first column from countData.
```{r}
# we need to remove the odd first $length col so the countData and colData files match up
countData <- as.matrix(countData[,-1])
head(countData)
```

Q. Complete the code below to filter countData to exclude genes (i.e. rows) where we have 0 read count across all samples (i.e. columns).
Tip: What will rowSums() of countData return and how could you use it in this context?
```{r}
# Filter count data where you have 0 read count across all samples.

# countData = countData[rowSums(countData > 0), ]

# head(countData)
to.rm.ind <-rowSums(countData)==0
countData<-countData[!to.rm.ind,]
nrow(countData)
```

## (1) DESeq

```{r}
dds = DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
dds = DESeq(dds)
```

```{r}
dds
res = results(dds)
```

Q. Call the summary() function on your results to get a sense of how many genes are up or down-regulated at the default 0.1 p-value cutoff.
```{r}
summary(res)
```

Volcano plot:
```{r}
plot( res$log2FoldChange, -log(res$padj) )
```

Q. Improve this plot by completing the below code, which adds color and axis labels.
```{r}
# Make a color vector for all genes
mycols <- rep("gray", nrow(res) )
# Color red the genes with absolute fold change above 2
mycols[ abs(res$log2FoldChange) > 2 ] <- "red"
# Color blue those with adjusted p-value less than 0.01
# and absolute fold change more than 2
inds <- (res$padj<.01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"
plot( res$log2FoldChange, -log(res$padj), col=mycols, xlab="Log2(FoldChange)", ylab="-Log(P-value)" )
```

Q. Use the mapIDs() function multiple times to add SYMBOL, ENTREZID and GENENAME
annotation to our results by completing the code below.

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

columns(org.Hs.eg.db)

res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="SYMBOL",
                    multiVals="first")

res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="GENENAME",
                    multiVals="first")

head(res, 10)
```

Q. Finally for this section let's reorder these results by adjusted p-value and save them to a CSV file in your current project directory.
```{r}
res = res[order(res$pvalue),]
write.csv(res, file="lab14_deseq_results.csv")
```

## (2) Pathway analysis

```{r}
library(pathview)
library(gage)
library(gageData)
```

```{r}
data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```

The main gage() function requires a named vector of fold changes, where the names of the values are the Entrez gene IDs. Note that we used the mapIDs() function above to obtain Entrez gene IDs and we have the fold change results from DESeq2 analysis.
```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

Now, let’s run the gage pathway analysis.
```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
```

Look at the object returned from gage().
```{r}
attributes(keggres)
```

It is a list with three elements, "greater", "less" and "stats".
Lets look at the first few down (less) pathway results:
```{r}
# Look at the first few down (less) pathways
head(keggres$less)
```

Each keggres object is data matrix with gene sets as rows sorted by p-value.
The top "less/down" pathways is "Cell cycle" with the KEGG pathway identifier hsa04110.

Now, let's try out the pathview() function from the pathview package to make a pathway plot with our RNA-Seq expression results shown in color. 
To begin with lets manually supply a pathway.id (namely the first part of the "hsa04110 Cell cycle") that we could see from the print out above.

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
# A different PDF based output of the same data
pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)
```

Now, let's process our results a bit more to automagicaly pull out the top 5 upregulated pathways, then further process that just to get the pathway IDs needed by the pathview() function. We'll use these KEGG pathway IDs for pathview plotting below.
```{r}
# Focus on top 5 upregulated pathways here for demo purposes only
keggrespathways <- rownames(keggres$greater)[1:5]

# Extract the 8 character long IDs part of each string
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids
```
```{r}
pathview(gene.data=foldchanges, pathway.id=keggresids, species="hsa")
```

Q. Can you do the same procedure as above to plot the pathview figures for the top 5 down-
reguled pathways?
```{r}
# top 5 down-regulated
keggrespathways_down <- rownames(keggres$less)[1:5]
# Extract the 8 character long IDs part of each string
keggresids_down = substr(keggrespathways_down, start=1, stop=8)
keggresids_down
# pathway
pathview(gene.data=foldchanges, pathway.id=keggresids_down, species="hsa")
```

## (3) Gene ontology

We can also do a similar procedure with gene ontology. Similar to above, go.sets.hs has all GO terms. go.subs.hs is a named list containing indexes for the BP, CC, and MF ontologies. Let’s focus on BP (a.k.a Biological Process) here.

```{r}
data(go.sets.hs)
data(go.subs.hs)

gobpsets = go.sets.hs[go.subs.hs$BP]
gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)
lapply(gobpres, head)
```

## (4) Reactome Analysis

Reactome is database consisting of biological molecules and their relation to pathways and processes. Reactome, such as many other tools, has an online software available (https://reactome.org/) and R package available (https://bioconductor.org/packages/release/bioc/html/ReactomePA.html).

Let's now conduct over-representation enrichment analysis and pathway-topology analysis with Reactome using the previous list of significant genes generated from our differential expression results above.

First, Using R, output the list of significant genes at the 0.05 level as a plain text file:
```{r}
sig_genes <- res[res$padj <= 0.05 & !is.na(res$padj), "symbol"]
print(paste("Total number of significant genes:", length(sig_genes)))
write.table(sig_genes, file="significant_genes.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)
```

Then, to perform pathway analysis online go to the Reactome website (https://reactome.org/PathwayBrowser/#TOOL=AT). Select “choose file” to upload your significant gene list. Then, select the parameters “Project to Humans”, then click “Analyze”.

Q: What pathway has the most significant “Entities p-value”? Do the most significant pathways listed match your previous KEGG results? What factors could cause differences between the two methods?

Reactome: Cell Cycle, mitotic entities p-value = 5.28E-4
KEGG: UP = caffeine metabolism, drug metabolism, purine metabolism 
KEGG: down = cell cycle, dna replication, rna transport

Yes, they are similar. 
Reactome is looking at all sig genes, regardless if they're up- or down-regulated. Reactome is also pulling from different sources than KEGG.

## (5) GO online (OPTIONAL) 

Q: What pathway has the most significant “Entities p-value”? Do the most significant pathways listed match your previous KEGG results?

- regulation of cell migration involved in sprouting angiogenesis
- platelet-derived growth factor receptor signaling pathway
- negative regulation of metaphase/anaphase transition of cell cycle

Q.  What factors could cause differences between the two methods?

Again, GO is looking for all sig genes, regardless if they're up- or down-regulated. GO annotations are also different from KEGG.

